src/graph/serializers/graphcontext_type.c

```diff
@@ -4,176 +4,33 @@
-	unsigned short relation_count = GraphContext_SchemaCount(gc, SCHEMA_EDGE);
-	RedisModule_SaveUnsigned(rdb, relation_count);
-
-	// Serialize unified edge schema.
-	RdbSaveDummySchema(rdb);
-
-	// Name of label X #relation schemas.
-	for(unsigned short i = 0; i < relation_count; i++) {
-		Schema *s = gc->relation_schemas[i];
-		RdbSaveSchema(rdb, s);
-	}
-
-	// Serialize graph object
-	RdbSaveGraph(rdb, gc);
-
-	// #Indices.
-	uint32_t index_count = gc->index_count;
-	RedisModule_SaveUnsigned(rdb, index_count);
-
-	// Serialize each index
-	_GraphContextType_SerializeIndicies(rdb, gc);
-
-	// Unlock.
-	Graph_ReleaseLock(gc->g);
-}
-
 void *GraphContextType_RdbLoad(RedisModuleIO *rdb, int encver) {
-	/* Format:
-	 * graph name
-	 * #node schemas
-	 * attribute mapping (in encver 4), or unified node schema
-	 * node schema X #node schemas
-	 * #relation schemas
-	 * filler bytes (in encver 4), or unified relation schema
-	 * relation schema X #relation schemas
-	 * graph object
-	 * #indices
-	 * (index label, index property) X #indices
-	 */
+	GraphContext *gc = NULL;
 
 	if(encver > GRAPHCONTEXT_TYPE_ENCODING_VERSION) {
 		// Not forward compatible.
 		printf("Failed loading Graph, RedisGraph version (%d) is not forward compatible.\n",
 			   REDISGRAPH_MODULE_VERSION);
-		return NULL;
-	}
-
-	// TODO can have different functions for different versions here if desired
-
-	GraphContext *gc = rm_calloc(1, sizeof(GraphContext));
-
-	// _tlsGCKey was created as part of module load.
-	pthread_setspecific(_tlsGCKey, gc);
-
-	// Graph name
-	gc->graph_name = RedisModule_LoadStringBuffer(rdb, NULL);
-
-	gc->g = Graph_New(GRAPH_DEFAULT_NODE_CAP, GRAPH_DEFAULT_EDGE_CAP);
-
-	// #Node schemas
-	uint32_t schema_count = RedisModule_LoadUnsigned(rdb);
-
-	// Initialize property mappings
-	gc->attributes = NewTrieMap();
-	gc->string_mapping = array_new(char *, 64);
-
-	// Load the full attribute mapping (or the attributes from
-	// the unified node schema, if encoding version is < 4)
-	RdbLoadAttributeKeys(rdb, gc);
-
-	// Load each node schema
-	gc->node_schemas = array_new(Schema *, schema_count);
-	for(uint32_t i = 0; i < schema_count; i ++) {
-		array_append(gc->node_schemas, RdbLoadSchema(rdb, SCHEMA_NODE));
-		Graph_AddLabel(gc->g);
-	}
-
-	// #Edge schemas
-	schema_count = RedisModule_LoadUnsigned(rdb);
-
-	// If encoding version is < 4, load the attributes from the
-	// unified edge schema, otherwise skip filler bytes.
-	RdbLoadAttributeKeys(rdb, gc);
-
-	// Load each edge schema
-	gc->relation_schemas = array_new(Schema *, schema_count);
-	for(uint32_t i = 0; i < schema_count; i ++) {
-		array_append(gc->relation_schemas, RdbLoadSchema(rdb, SCHEMA_EDGE));
-		Graph_AddRelationType(gc->g);
-	}
-
-	// Graph object.
-	RdbLoadGraph(rdb, gc);
-
-	// #Indices
-	// (index label, index property) X #indices
-	uint32_t index_count = RedisModule_LoadUnsigned(rdb);
-	for(uint32_t i = 0; i < index_count; i ++) {
-		RdbLoadIndex(rdb, gc);
+	} else {
```

Comment:
Broadly speaking, I like this new format for the serialization files! Shouldn't we have a check here to switch deserialization routines for encver == current-1 and an error on encver < current-1, though? Right now this doesn't have forward compatibility.

Reply:
We sure do, added.